File: VB/WebSite/App_Code/CustomResources.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel

<Serializable> _
Public Class CustomResourceList
	Inherits BindingList(Of CustomResource)
	Public Sub AddRange(ByVal resources As CustomResourceList)
		For Each customResource As CustomResource In resources
			Me.Add(customResource)
		Next customResource
	End Sub
	Public Function GetResourceIndex(ByVal resourceId As Object) As Integer
		For i As Integer = 0 To Count - 1
			If Me(i).Id Is resourceId Then
				Return i
			End If
		Next i
		Return -1
	End Function
End Class
Public Class CustomResourceDataSource
	Private resources_Renamed As CustomResourceList
	Public Sub New(ByVal resources As CustomResourceList)
		If resources Is Nothing Then
			DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("resources")
		End If
		Me.resources_Renamed = resources
	End Sub
	Public Sub New()
		Me.New(New CustomResourceList())
	End Sub
	Public Property Resources() As CustomResourceList
		Get
			Return resources_Renamed
		End Get
		Set(ByVal value As CustomResourceList)
			resources_Renamed = value
		End Set
	End Property

	#Region "ObjectDataSource methods"
	Public Sub InsertMethodHandler(ByVal customResource As CustomResource)
		Resources.Add(customResource)
	End Sub
	Public Sub DeleteMethodHandler(ByVal customResource As CustomResource)
		Dim resourceIndex As Integer = Resources.GetResourceIndex(customResource.Id)
		If resourceIndex >= 0 Then
			Resources.RemoveAt(resourceIndex)
		End If
	End Sub
	Public Sub UpdateMethodHandler(ByVal customResource As CustomResource)
		Dim resourceIndex As Integer = Resources.GetResourceIndex(customResource.Id)
		If resourceIndex >= 0 Then
			Resources.RemoveAt(resourceIndex)
			Resources.Insert(resourceIndex, customResource)
		End If
	End Sub
	Public Function SelectMethodHandler() As IEnumerable
		Dim result As New CustomResourceList()
		result.AddRange(Resources)
		Return result
	End Function
	#End Region
End Class

<Serializable> _
Public Class CustomResource
	Private id_Renamed As Object
	Private caption_Renamed As String

	Public Sub New()
	End Sub

	Public Property Id() As Object
		Get
			Return id_Renamed
		End Get
		Set(ByVal value As Object)
			id_Renamed = value
		End Set
	End Property
	Public Property Caption() As String
		Get
			Return caption_Renamed
		End Get
		Set(ByVal value As String)
			caption_Renamed = value
		End Set
	End Property
End Class

File: CS/WebSite/App_Code/CustomResources.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;

[Serializable]
public class CustomResourceList : BindingList<CustomResource> {
	public void AddRange(CustomResourceList resources) {
		foreach(CustomResource customResource in resources)
			this.Add(customResource);
	}
	public int GetResourceIndex(object resourceId) {
		for(int i = 0; i < Count; i++)
			if(this[i].Id == resourceId)
				return i;
		return -1;
	}
}
public class CustomResourceDataSource {
	CustomResourceList resources;
	public CustomResourceDataSource(CustomResourceList resources) {
		if(resources == null)
			DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("resources");
		this.resources = resources;
	}
	public CustomResourceDataSource()
		: this(new CustomResourceList()) {
	}
	public CustomResourceList Resources { get { return resources; } set { resources = value; } }

	#region ObjectDataSource methods
	public void InsertMethodHandler(CustomResource customResource) {
		Resources.Add(customResource);
	}
	public void DeleteMethodHandler(CustomResource customResource) {
		int resourceIndex = Resources.GetResourceIndex(customResource.Id);
		if(resourceIndex >= 0)
			Resources.RemoveAt(resourceIndex);
	}
	public void UpdateMethodHandler(CustomResource customResource) {
		int resourceIndex = Resources.GetResourceIndex(customResource.Id);
		if(resourceIndex >= 0) {
			Resources.RemoveAt(resourceIndex);
			Resources.Insert(resourceIndex, customResource);
		}
	}
	public IEnumerable SelectMethodHandler() {
		CustomResourceList result = new CustomResourceList();
		result.AddRange(Resources);
		return result;
	}
	#endregion
}

[Serializable]
public class CustomResource {
	object id;
	string caption;

	public CustomResource() {
	}

	public object Id { get { return id; } set { id = value; } }
	public string Caption { get { return caption; } set { caption = value; } }
}